### PR TITLE
refactor(resource): collapse topology taxonomy + close isolation defects

### PR DESCRIPTION
**Pull request by vanyastaff**
## Summary

`nebula-resource` no longer fans a 2-axis decision into a 5-topology God-object. The `Service` / `Transport` / `Exclusive` topologies collapse into one parameterized `Bounded` runtime behind a sealed `Cap` typestate, folding ~35 hand-copied Manager methods and 5 byte-identical acquire pipelines into one generic path. Riding the same seam, three latent isolation/correctness defects the collapse exposed are closed: the cross-tenant identity barrier, the pool revoke-recycle TOCTOU, and swallowed release-hook errors. Externally observable behavior is preserved except those explicitly-listed fixes; the whole workspace is green at every commit (expand-contract, no shim).

These ship together because the security fixes are only safely expressible *through* the collapsed seam — a 5-way duplicated barrier cannot be fixed in one place until it is one place.

## What changed

| Area | Before | After |
|---|---|---|
| Topology model | 5 traits + 5 runtimes + 5 config modules | `Pool` / `Resident` / `Bounded<Cap>` (`Unbounded` / `Capped<N>` / `Exclusive`) |
| Acquire pipeline | 5 byte-identical `run_*_acquire` + dead `_ => expected X` arms | one generic `run_acquire`; one registry walk (typed lookup is a downcast, not a second DashMap walk) |
| Cross-tenant identity | `DefaultHasher` → `u64`; a collision silently merged tenants | structural ordered `(slot, credential)` key with derived `Eq`/`Hash` — collision impossible by construction |
| Pool revoke | revoke hook walked only currently-idle entries; recycle / in-flight-create / warmup / maintenance re-admitted a revoked credential | per-`PoolRuntime` revoke epoch snapshotted at checkout, re-checked at all six return-to-idle paths; destroy-not-admit before the hook |
| Release-hook errors | `let _ =`-swallowed | observed (tracing + a release-error metric); a failed Exclusive reset destroys the instance and still returns the permit |
| Register surface | ~17 shorthands + a 3-deep delegation chain + 4 `too_many_arguments` allows | one `register(RegistrationSpec)` funnel |
| Registry pinned lookup | handled an impossible `Ambiguous` with a fabricated error | 2-variant `Found`/`NotFound` (impossible state unrepresentable); identity-agnostic paths keep their real fail-closed `Ambiguous` |
| Dead surface | `recovery/{group,watchdog}`, identity-agnostic acquire tier, detach sentinel + `Deref` panic | deleted; post-detach guard is a borrow-checker error, not a runtime panic |

## Design decisions

- **Sealed `Cap` typestate, not a runtime discriminant.** Invalid combinations (a tracked lease with no release hook, an exclusive resource with no reset) are compile errors via a `BoundedRelease` sub-trait bound, not the old runtime `TOKEN_MODE ==` branch. `RESET_ON_RELEASE` (true only for `Exclusive`) gates the destroy-on-failed-reset so a single-session transport's shared runtime is never destroyed.
- **Structural identity, not a digest.** A digest keeps a collidable space; the key is the exact ordered `(slot, credential)` set. The engine receives the identity from the single construction site and propagates it rather than recomputing — there is no second site to drift. A cross-crate test asserts engine and resource derive byte-identical keys and that distinct credentials never alias.
- **Revoke epoch snapshotted at checkout, not loaded at release.** A release-time load reads the post-revoke value on pre-revoke entries too and cannot distinguish them; the epoch is stamped on the `PoolEntry` at create/checkout and the bump is synchronous before the hook so a dropped drain-tail future cannot un-fence it.
- **Expand-contract, no shim.** Add the new surface, migrate every in-tree consumer (including the engine's structural-identity persistence and rotation fan-out), then delete the old surface in one atomic commit gated by `cargo check -p` for every consumer crate. A U1↔U11 golden-replay oracle asserts the post-collapse `Bounded` runtime emits byte-identical observable outcomes to three committed pre-collapse fixtures.

The architectural rationale (why 5→3, why RCU was rejected for `SlotCell`, the two-phase revoke invariant) is consolidated once in the `manager` module doc as the durable record — no ADR.

## Behavior carve-out (intentional, recorded)

A former-`Service` resource's `reload_config` now returns `SwappedImmediately` instead of `PendingDrain`, because it is `TopologyRuntime::Bounded` post-fold. This is a no-op-preserving relabel: that path never drained or rebuilt the live runtime either way (the real fix is the separately-tracked `reload_config` redesign). The reload-deferral test is re-pinned to the post-fold value so the change is locked, not silent.

## Test summary

Per-crate `cargo check --workspace --all-targets`, clippy `-D warnings` (default + `rotation`), nextest (resource + engine, default + `rotation`), doctests, `rustdoc -D warnings`, and the golden-replay oracle are all green. Characterization-first: the R15/R16/R17 nets were RED before their fixes and are GREEN after, with 0 unexpected skips. A 12-persona code review (security / adversarial / performance / api-contract returned clean) caught one P1 regression (S4 destroy firing for `Capped<1>`), now fixed with a RED-proof test.

## Residual Review Findings

Non-blocking follow-ups from the multi-agent review and the deferred-bug ledger, filed as tracked issues so nothing is silently inherited:

- [vanyastaff/nebula#712](https://github.com/vanyastaff/nebula/issues/712) — `reload_config` no-drain redesign (MED-HIGH)
- [vanyastaff/nebula#713](https://github.com/vanyastaff/nebula/issues/713) — pool `CreateGuard` cancel-drop leak (MED)
- [vanyastaff/nebula#714](https://github.com/vanyastaff/nebula/issues/714) — resident recreate take()+destroy window (MED)
- [vanyastaff/nebula#715](https://github.com/vanyastaff/nebula/issues/715) — `graceful_shutdown` phase-4 detached workers (LOW)
- [vanyastaff/nebula#716](https://github.com/vanyastaff/nebula/issues/716) — `RecoveryTicket` panicked-probe counts as attempt (LOW)
- [vanyastaff/nebula#717](https://github.com/vanyastaff/nebula/issues/717) — separable acquire-path micro-allocation folds (perf, LOW)
- [vanyastaff/nebula#718](https://github.com/vanyastaff/nebula/issues/718) — cross-crate dedup / layer-placement follow-ups
- [vanyastaff/nebula#719](https://github.com/vanyastaff/nebula/issues/719) — fold the identity-agnostic vs identity-pinned method axis (Manager LoC, LOW)
- [vanyastaff/nebula#720](https://github.com/vanyastaff/nebula/issues/720) — `SlotIdentity` `#[non_exhaustive]` + `Capped<0>` guard (LOW)
- [vanyastaff/nebula#721](https://github.com/vanyastaff/nebula/issues/721) — delete dead `pub Registry::get_typed_for` orphaned by the collapse (LOW)
- [vanyastaff/nebula#722](https://github.com/vanyastaff/nebula/issues/722) — pool checkout-epoch-guard e2e test gap + `acquire_dispatch` 3-hop flatten (LOW)

The structural de-spaghettification goals are met and type-enforced against regression; the literal "~800-line `manager/mod.rs`" target is honestly not met (2552 total, ~1211 code, ~1224 the centralized canonical doc) and is tracked as [#719](https://github.com/vanyastaff/nebula/issues/719).

---

[Compound Engineering](https://github.com/EveryInc/compound-engineering-plugin)
Claude Code

## Summary by CodeRabbit

* **New Features**
  * Collision-free structural slot identities for stronger multi-tenant isolation.
  * Bounded runtime unifying capped/exclusive/session semantics.
  * Resource release-error metric to surface release-hook failures.

* **API Changes**
  * Registration consolidated to a single RegistrationSpec-based funnel and slot-identity wiring.
  * Topology surface collapsed to Pool, Resident, and parameterized Bounded.

* **Bug Fixes**
  * Pool revoke fencing to prevent revoked credentials re-entering idle pools.

* **Documentation**
  * README, docs, examples, and tests updated for the new model.

[Review Change Stack](https://app.coderabbit.ai/change-stack/vanyastaff/nebula/pull/723?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: crates/resource/src/runtime/bounded.rs [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+        // session rather than resetting the shared runtime, and the next
+        // acquirer needs that runtime — destroying it would be a
+        // regression. (The failed-release error is still observed above and
+        // the permit is still returned below for every cap.) `destroy`
+        // consumes an owned `R::Runtime`; the lease was minted by cloning
+        // the shared runtime, so a clone is the matching instance to tear
+        // down (the shared `Arc` stays pristine for the next acquire).
+        if R::Cap::RESET_ON_RELEASE {
+            let instance = (*runtime).clone();
+            if let Err(de) = resource.destroy(instance).await {
```
[the remark this answers is not in the document]
Comment (vanyastaff) [reply crates/resource/src/runtime/bounded.rs:476]: > destroy on (*runtime).clone() but leaves BoundedRuntime.runtime unchanged, so future acquires still use the same in-memory runtime instance

Addressed (P1): a failed Exclusive reset now latches an `AtomicBool` poison flag (set before the best-effort destroy, gated on `Cap::RESET_ON_RELEASE` so Exclusive-only). Every subsequent `acquire` checks it and returns `Error::permanent` before minting a lease instead of serving the poisoned shallow-clone; the permit is still returned (no semaphore deadlock). New test `bounded_exclusive_failed_reset_poisons_runtime_next_acquire_fails_closed` proves it; `Capped<N>`/`Unbounded` unaffected.

File: crates/resource/src/runtime/bounded.rs [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+    pub fn new(resource: &R, runtime: R::Runtime, config: Config) -> Self {
+        let runtime = Arc::new(runtime);
+        let semaphore = R::Cap::PERMITS.map(|n| Arc::new(Semaphore::new(n)));
+        let keepalive = config.keepalive_interval.map(|interval| {
+            let resource = resource.clone();
+            let runtime = Arc::clone(&runtime);
+            tokio::spawn(async move {
+                let mut ticker = time::interval(interval);
+                ticker.set_missed_tick_behavior(time::MissedTickBehavior::Skip);
+                // First tick fires immediately; skip it so the first
```
[the remark this answers is not in the document]
Comment (vanyastaff) [reply crates/resource/src/runtime/bounded.rs]: > BoundedRuntime::new spawns the keepalive task via tokio::spawn from a synchronous constructor ... this panics at runtime (library panic)

Addressed: `new` no longer spawns; the keepalive task starts lazily on the first async `acquire` (ambient runtime guaranteed) behind a `std::sync::Once`, with the `JoinHandle` parked in a `Mutex<Option<_>>` so `Drop` still aborts it (no leak). No off-runtime panic; keepalive still fires for Transport-shaped caps (test updated + an off-runtime no-panic test added).

File: crates/resource/src/registry.rs [hunk 3]
```diff
@@ -445,54 +530,49 @@ impl Registry {
...
+                    if slot_identity.is_unbound()
+                        && Self::scope_has_cred_bound_rows_without_unbound(&entries, &level)
+                    {
+                        return PinnedLookup::NotFound;
+                    }
+                    continue;
+                },
+            }
         }
+        PinnedLookup::NotFound
```
[the remark this answers is not in the document]
Comment (vanyastaff) [reply crates/resource/src/registry.rs:609]: > type_index only narrows this to a ResourceKey ... can return a sibling row of another type at the requested scope+identity and fail later on downcast instead of returning NotFound

Addressed: threaded an `Option<TypeId>` concrete-type constraint through `find_pinned_at_exact_scope`/`find_pinned_in_entries`/`scope_has_cred_bound_rows_without_unbound`; typed entry points pass `TypeId::of::<ManagedResource<R>>()` so a sibling-typed row under the shared key is skipped and the scope walk continues. Pinned path stays 2-variant; agnostic path keeps fail-closed Ambiguous. Test `pinned_finder_skips_sibling_typed_row_under_shared_key` added.

File: crates/resource/src/registry.rs [hunk 4]
```diff
@@ -691,6 +767,57 @@ impl Registry {
...
+        } else if *scope != ScopeLevel::Global
+            && entries.iter().any(|e| e.scope == ScopeLevel::Global)
+        {
+            ScopeLevel::Global
+        } else {
+            return PinnedFind::NotFound;
+        };
+
+        Self::find_pinned_at_exact_scope(entries, &effective_scope, want_identity)
+    }
```
[the remark this answers is not in the document]
Comment (vanyastaff) [reply crates/resource/src/registry.rs:967]: > find_pinned_in_entries chooses the effective scope before it checks want_identity ... Direct pinned lookup disagrees with acquire routing

Addressed: `find_pinned_in_entries` is now identity-aware — exact `scope` with `want_identity` first, then on `NotFound` (scope != Global) retry at `Global` with the same identity+type pin, instead of selecting effective scope by "any entry exists". Direct pinned lookup now agrees with acquire routing; fail-closed preserved (Global fallback only matches an exact `slot_identity` + type). Two regression tests added.

File: crates/resource/src/manager/execute.rs [hunk 5]
```diff
@@ -42,23 +42,9 @@ where
...
-        )));
-    }
-    Ok(())
-}
+// Pool `(min_size, max_size)` sanity (#390) is enforced structurally by
+// `PoolRuntime::new`'s construction-time asserts, which every caller must
+// invoke to build a `TopologyRuntime::Pool`. The former register-time
+// soft-`Err` re-check (used only by the deleted `register_pooled[_with]`
+// shorthands, which took the raw config *before* building the runtime) is
+// gone — the invariant is now unrepresentable rather than re-validated.
```
[the remark this answers is not in the document]
Comment (vanyastaff) [reply crates/resource/src/manager/execute.rs]: > Line 46 shifts (min_size, max_size) validation to constructor assert behavior, which can turn invalid user config into process panic instead of a typed registration error

Addressed: added `PoolRuntime::try_new(...) -> Result<Self, Error>` performing the #390 (min/max) check and returning `Error::permanent`; the engine/JSON registration path now uses it so invalid topology fails as a typed registration error, not a process panic. `PoolRuntime::new` (assert) retained only for compile-time-known callers. Ported the deleted `register_pooled_rejects_*` coverage onto `try_new`.

File: crates/resource/tests/basic_integration.rs [hunk 6]
```diff
@@ -2638,17 +2883,16 @@ async fn acquire_retries_on_transient_failure() {
...
+        .register(RegistrationSpec {
+            resource: resource.clone(),
+            config: test_config(),
+            scope: ScopeLevel::Global,
+            slot_identity: SlotIdentity::Unbound,
+            topology: TopologyRuntime::Resident(resident_rt),
+            acquire: Manager::erased_acquire_resident_for::<ResidentTestResource>(),
+            resilience: Some(resilience),
+            recovery_gate: None,
+        })
```
[the remark this answers is not in the document]
Comment (vanyastaff) [reply crates/resource/tests/basic_integration.rs:2917]: > These specs register FailingResidentResource / PermanentFailResource, but the stored hook is Manager::erased_acquire_resident_for::<ResidentTestResource>()

Addressed: every flagged `RegistrationSpec.acquire` now uses the hook matching the registered resource — `erased_acquire_resident_for::<FailingResidentResource>()` / `::<PermanentFailResource>()` across all listed ranges (and an identical mismatch at the herd test).

File: crates/resource/tests/basic_integration.rs [hunk 7]
```diff
@@ -3711,17 +3995,16 @@ async fn reload_config_swaps_config_and_bumps_generation() {
...
-            None,
-        )
+            config: ReloadConfig::new(1),
+            scope: ScopeLevel::Global,
+            slot_identity: SlotIdentity::Unbound,
+            topology: TopologyRuntime::Pool(pool_rt),
+            acquire: Manager::erased_acquire_pooled_for::<PoolTestResource>(),
+            resilience: None,
+            recovery_gate: None,
+        })
```
[the remark this answers is not in the document]
Comment (vanyastaff) [reply crates/resource/tests/basic_integration.rs:4051]: > These registrations pair ReloadPoolResource with Manager::erased_acquire_pooled_for::<PoolTestResource>(), so any path consuming the stored hook downcasts to the wrong managed type

Addressed: all five reload-test registrations now use `erased_acquire_pooled_for::<ReloadPoolResource>()` to match the registered resource.

File: crates/resource/tests/basic_integration.rs [hunk 8]
```diff
@@ -3257,7 +3501,12 @@ async fn pool_recycle_drop_destroys_entry() {
         .expect("acquire should succeed");
 
     drop(handle);
-    tokio::time::sleep(std::time::Duration::from_millis(50)).await;
+    // recycle=Drop destroys the instance — wait for `destroy` to run (idle
+    // stays 0 throughout, so the destroy counter is the settle signal).
+    poll_until(std::time::Duration::from_secs(2), || {
+        resource.destroy_counter.load(Ordering::Relaxed) >= 1
+    })
+    .await;
```
[the remark this answers is not in the document]
Comment (vanyastaff) [reply crates/resource/tests/basic_integration.rs]: > poll_until returns a boolean, but this call ignores it and the only follow-up assertion is idle_count == 0, which also holds if destroy never ran

Addressed: wrapped the `poll_until(...)` in `pool_recycle_drop_destroys_entry` with `assert!(..., "destroy never ran for recycle=Drop")` so the drop path is proven before the `idle_count == 0` assertion.

File: crates/resource/src/slot.rs [hunk 9]
```diff
@@ -259,4 +259,104 @@ mod tests {
...
+                        assert!(
+                            observed_gen >= 1,
+                            "a published entry always has generation >= 1"
+                        );
+                        // `val.0` was the `g_next` of whichever store
+                        // published this entry; that store's published
+                        // generation is `observed_gen`. They must be
+                        // consistent: the entry is a single immutable unit.
+                        let _ = val.0;
+                    }
```
[the remark this answers is not in the document]
Comment (vanyastaff) [reply crates/resource/src/slot.rs]: > it never asserts any relationship between observed_gen and val.0 (it just does let _ = val.0;) ... can pass even if load_versioned returns mismatched pairs

Addressed: added an in-module `store_stamped` that bumps generation then derives the value from it (value == generation via the same single ArcSwapOption publish); the test now asserts `assert_eq!(u64::from(val.0), observed_gen)` on every `load_versioned`. Mutation-verified: injecting `generation + 1` makes it fail where the old `let _ =` passed.

File: crates/resource/macros/src/resource_attrs.rs [hunk 10]
```diff
@@ -99,9 +100,10 @@ impl ResourceAttrs {
...
             "resident" => "Resident",
-            "service" => "Service",
-            "transport" => "Transport",
-            "exclusive" => "Exclusive",
+            "bounded" => "Bounded",
+            // guard-justified: every reachable value is one of the three
+            // strings validated by `parse()` above; this arm is
+            // unreachable by construction.
             _ => unreachable!("topology validated in parse()"),
         };
```
[the remark this answers is not in the document]
Comment (vanyastaff) [reply crates/resource/macros/src/resource_attrs.rs:121]: > Prefer parsing topology into a small internal enum in parse() and storing that in ResourceAttrs, so topology_ident() can be an exhaustive match with no panic paths

Addressed: added an internal `Topology { Pool, Resident, Bounded }` enum parsed once in `parse()` (where the unknown-value `syn::Error` is already emitted); `topology_ident()` is now an exhaustive match with no `unreachable!()`/panic arm. Validation diagnostics unchanged, so the trybuild stderr fixtures still match.

File: crates/resource/src/topology/bounded.rs [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+/// former `Transport` topology: the runtime gates concurrency with a
+/// `Semaphore(N)` and runs `release_one` (token return / session close)
+/// on lease drop. For the transport shape, `N` is `max_sessions`.
+#[derive(Debug, Clone, Copy)]
+pub struct Capped<const N: usize>;
+
+impl<const N: usize> sealed::Sealed for Capped<N> {}
+impl<const N: usize> CapMarker for Capped<N> {
+    const PERMITS: Option<usize> = Some(N);
+    const RELEASE_REQUIRED: bool = true;
```
[the remark this answers is not in the document]
Comment (vanyastaff) [reply crates/resource/src/topology/bounded.rs:143]: > Capped<0> currently creates a permanently unavailable bounded resource ... This should be invalid by construction

Addressed (fixed-differently): `BoundedRuntime::new` carries a firing `const { assert!(!matches!(R::Cap::PERMITS, Some(0)), ...) }` so building with `Cap = Capped<0>` is a hard compile error. Proven via a `compile_fail` doctest rather than trybuild — trybuild runs `cargo check`, under which a post-monomorphization const assert in a generic fn is not evaluated; the type-check alternative needs nightly `generic_const_exprs`. This supersedes deferred issue #720 (now closed).

File: crates/resource/tests/golden/mod.rs [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+        match c {
+            '"' => out.push_str("\\\""),
+            '\\' => out.push_str("\\\\"),
+            '\n' => out.push_str("\\n"),
+            '\r' => out.push_str("\\r"),
+            '\t' => out.push_str("\\t"),
+            other => out.push(other),
+        }
+    }
+    out
```
[the remark this answers is not in the document]
Comment (vanyastaff) [reply crates/resource/tests/golden/mod.rs:147]: > escape currently leaves some control bytes unescaped, so the emitted fixture can become invalid JSON for edge-case event content

Addressed: `escape` now emits `\b`/`\f` for U+0008/U+000C and `\u00XX` for any remaining char <= 0x1F (short escapes for `\n \r \t \" \` retained). AE1/AE2/AE3 goldens verified byte-equal — golden_replay/bounded_fold_behavior pass with no fixture diff.

File: crates/resource/tests/manager_acquire_for.rs [hunk 13]
```diff
@@ -1,21 +1,22 @@
...
-//! refactors of the resident pattern proven in `dedup_slot_identity.rs`
-//! (same `lookup_for_acquire_with` → shared `run_*_acquire`), so they are
-//! not re-mocked here. `refresh_slot_for` / `revoke_slot_for` (the ports the
-//! engine rotation fan-out drives) are covered directly.
+//! Pooled is covered end-to-end (the most common topology); the folded
+//! `Bounded` `_for_identity` methods are line-identical refactors of the
+//! resident pattern proven in `dedup_slot_identity.rs` (same
+//! `lookup_for_acquire_with` → shared `run_acquire`), so they are not
+//! re-mocked here. `refresh_slot_for_identity` / `revoke_slot_for_identity`
+//! (the ports the engine rotation fan-out drives) are covered directly.
```
[the remark this answers is not in the document]
Comment (vanyastaff) [reply crates/resource/tests/manager_acquire_for.rs:19]: > without one acquire_bounded_for_identity / revoke_slot_for_identity two-tenant case, a bounded-specific routing regression can slip through

Addressed: added `acquire_bounded_for_resolves_and_revokes_only_the_pinned_row` — a `Bounded` (`Capped<2>`) two-tenant regression mirroring the pooled/resident pattern: two rows under one (key, scope) with distinct structural identities, each `acquire_bounded_for_identity` resolves only its own runtime, agnostic acquire fail-closed, `revoke_slot_for_identity` taints only tenant A while B keeps serving.

File: docs/plans/2026-05-18-004-refactor-nebula-resource-topology-collapse-plan.md [hunk 14]
```diff
@@ -0,0 +1,542 @@
...
+   → cfg = managed.config()  [hoisted ONCE, &cfg into closure — not per retry]
+   → execute_with_resilience(&managed.resilience)               [no .clone()]
+   → topology.acquire(cx)    [enum dispatch; Bounded folds Service/Transport/Exclusive]
+   → settle gate / record (broadcast only if receiver_count>0)
+   → with_drain_tracker
+
+revoke (R16 fix): drain MUST account for in-flight whose recycle-to-idle
+  is still queued on the ReleaseQueue — close the gap so a revoked
+  credential cannot re-enter idle after the hook walked it.
+```
```
[the remark this answers is not in the document]
Comment (vanyastaff) [reply docs/plans/2026-05-18-004-refactor-nebula-resource-topology-collapse-plan.md]: > The code fence starting at Line 124 is unlabeled, which triggers markdownlint MD040

Addressed: added a `text` language tag to the opening fence of the TopologyRuntime/acquire-revoke ASCII sketch. Block contents unchanged.

File: docs/plans/2026-05-18-004-refactor-nebula-resource-topology-collapse-plan.md [hunk 15]
```diff
@@ -0,0 +1,542 @@
...
+
+### Deferred to Implementation
+- [Affects U1][Needs codebase research — NOT external; narrowed] Is concurrent same-slot `SlotCell::store` reachable? Closed by research: (a) arc-swap plain `store`/`swap` has no inter-writer ordering (confirmed) — so non-monotonicity is real *only* under concurrent same-slot store; (b) the engine rotation fan-out (`resource_fanout.rs`) dispatches per-resource-**concurrently** for ONE rotation event via `join_all` — distinct resources = distinct `SlotCell`s, **no same-slot contention at the fan-out layer**. Residual (the only remaining unknown): are two rotation *events* for the **same credential** serialized upstream (rotation driver/ledger)? If yes → single-writer-per-slot holds, document the invariant (expected). If no → real non-monotonicity. Guarded by the U1 concurrency test; resolve the upstream-serialization trace at U1, not here (different crate's driver internals).
+- [Affects U4][Technical] Keep `service/transport/exclusive` as accepted `topology=` catalog aliases or collapse to one string (cosmetic — macro emits only the informational const).
```
[the remark this answers is not in the document]
Comment (vanyastaff) [reply docs/plans/2026-05-18-004-refactor-nebula-resource-topology-collapse-plan.md]: > [Needs codebase research — NOT external; narrowed] and [Technical] are parsed as reference labels but have no definitions (MD052)

Addressed: converted the bracketed Deferred-to-Implementation annotations to parenthesized plain text (`[Affects U1][Needs codebase research — NOT external; narrowed]` → `(Affects U1)(Needs codebase research — NOT external; narrowed)`, same for `(Affects U4)(Technical)`). Wording unchanged; brackets no longer parse as undefined link refs.

**Review by vanyastaff — COMMENTED** [review]
Reviewed clean PR head `7521ade7612d27a0776b032827bf2cb810e54c6f`.

Findings I would block on before merge:

- `crates/engine/src/engine.rs:214` / `:667`: P1: `resource_slot_identities` is keyed only by `ResourceKey`, so registering two tenants/rows for the same resource key overwrites the first identity with the last one. Later executions snapshot that process-wide value and can acquire the wrong tenant row. Key this by the actual activation row, at least `(ResourceKey, ScopeLevel, SlotIdentity)`, or store the per-execution resolved bindings instead of a global per-key map.
- `crates/engine/src/resource_accessor.rs:69`: P1: missing slot identity falls back to `SlotIdentity::Unbound`. For credential-bound resources, a stale/absent identity can silently acquire an unbound fallback row if one exists. Make missing identity fail closed for credential-bound keys.
- `crates/resource/src/registry.rs:542`: P1: typed lookup resolves `TypeId -> ResourceKey`, then row search matches only `(scope, slot_identity)`. If sibling concrete types share a key, it can return the wrong `ManagedResource` and downcast to `NotFound` even when the right typed row exists later. Filter rows by `managed_type_id()` before returning a pinned hit.
- `crates/resource/src/registry.rs:804`: P1: `find_pinned_in_entries` picks the requested scope if any row exists there, even when none match `want_identity`, so a matching `Global` row is never considered. Choose exact scope only when it contains the requested identity; otherwise allow the global fallback.
- `crates/resource/src/runtime/bounded.rs:269`: P1: failed `Exclusive` reset destroys `(*runtime).clone()` but leaves `BoundedRuntime.runtime` installed. For shallow-clone runtimes, the next acquire still uses the poisoned shared instance. Quarantine/replace the stored runtime, or mark it failed so future acquires recreate instead of reusing it.
- `crates/resource/src/runtime/pool.rs:137`: P2: `PoolRuntime::new` panics on invalid `min_size/max_size` via `assert!`. This is public library construction fed by config-shaped input, so it should be a typed error. Add `try_new` or move validation into the fallible registration path.

Verification run locally on a clean detached worktree: `cargo check -p nebula-resource --all-targets` and `cargo check -p nebula-engine --all-targets --features rotation` both pass. I did not run full nextest.

**Comment by vanyastaff** [comment]
> CodeRabbit nitpick (guard.rs ~201-218): "detach() now skips the Drop branch that explicitly extracts OwnedSemaphorePermit ... a future refactor can leak bounded/exclusive capacity without tripping the current panic-path coverage."

Addressed in f76d906a: added `guard::tests::detach_guarded_returns_permit_to_semaphore` — builds a Guarded `ResourceGuard` holding the only permit of `Semaphore::new(1)`, asserts the semaphore is exhausted while held, calls `detach()`, then asserts the permit is reclaimed (`available_permits()` back to 1, re-acquirable) while `detach()` still yields the lease. A future refactor that leaks the implicitly-dropped permit now fails this test. nebula-resource suite 328/328.

(The other non-thread item — the `register_and_bind` register-then-bind fan-out window flagged outside the diff range — is a pre-existing, documented deferred contract, not introduced by this PR; it is tracked in the crate's deferred-bug ledger and is intentionally out of scope here.)

File: crates/resource/src/registry.rs [hunk 16]
```diff
@@ -691,6 +817,106 @@ impl Registry {
...
+    /// `NotFound`/ancestor-row and a spurious `NotFound` masking a
+    /// correctly-typed row at a later scope. Untyped callers
+    /// ([`get_for`](Self::get_for)) pass `None`: they hand back the erased
+    /// `Arc` without downcasting, so no concrete type is implied.
+    fn entry_type_matches(entry: &RegistryEntry, concrete_type: Option<TypeId>) -> bool {
+        match concrete_type {
+            Some(tid) => entry.managed.managed_type_id() == tid,
+            None => true,
+        }
+    }
```
[the remark this answers is not in the document]
Comment (vanyastaff) [reply crates/resource/src/registry.rs:889]: Fixed in c5eb64dc — threaded `Option<TypeId>` through `find_at_exact_scope`/`find_in_entries` (effective-scope resolution included) via a shared `get_inner`; `get_typed::<R>` and `get_typed_for_acquire_scope::<R>` now pass `Some(TypeId::of::<ManagedResource<R>>())`, erased callers keep `None`. Added two FakeA/FakeB shared-key regression tests (skip-and-fall-through + ambiguity-not-inflated).

File: crates/resource/tests/basic_integration.rs [hunk 17]
```diff
@@ -1488,8 +1603,11 @@ async fn pool_detach_removes_from_pool() {
     let lease = handle.detach();
     assert!(lease.is_some(), "guarded handle detach should return Some");
 
-    // Give time for any potential (but should-not-happen) release processing.
-    tokio::time::sleep(std::time::Duration::from_millis(50)).await;
+    // `detach` disarms the release callback synchronously, so nothing can
+    // ever be submitted to the queue. A single scheduler yield lets any
+    // (erroneously) spawned release task run before we assert — a
+    // deterministic turn, not a wall-clock guess.
+    tokio::task::yield_now().await;
```
[the remark this answers is not in the document]
Comment (vanyastaff) [reply crates/resource/tests/basic_integration.rs]: Fixed in c5eb64dc — replaced `yield_now()` with `drop(rq)` + `ReleaseQueue::shutdown(rq_handle).await` *before* the `idle_count == 0` assert, so any erroneously-enqueued release has run by then ("never", not "not yet").

**Comment by vanyastaff** [comment]
@coderabbitai review last commits

**Comment by vanyastaff** [comment]
Addressed in `0aa5709a` (this PR): added `typed_lookup_skips_sibling_type_and_falls_through_to_global` in `crates/resource/tests/dedup_slot_identity.rs`, mirroring the acquire-path guard. Mutation-verified — reverting `Registry::get_typed` to the unfiltered `self.get(&key, scope)` fails **only** this new test; the acquire-path E2E and the other 336 stay green, confirming it closes the exact uncovered vector.

One correction to the suggestion: `Manager::get_any` is the type-erased entrypoint (`registry.get(key, scope)`, no `TypeId`) and never exercises this filter. The real public typed delegator is `Manager::lookup::<R>` → `Registry::get_typed::<R>` → `get_inner(.., Some(type_id))`; the test targets `lookup` and the doc records the correction so it reads accurately standalone. A positive control asserts the sibling type still resolves its own org row (filter is type-directed, not a blanket scope skip).

Note: unrelated audit-hardening work that was briefly stacked on this branch has been split to #724 to keep this PR's review scope to the detach/concrete-type-filter change.
